internal/services/v1/watch_test.go

```diff
@@ -0,0 +1,206 @@
+		{
+			name:         "unfiltered watch",
+			expectedCode: codes.OK,
+			mutations: []*v1.RelationshipUpdate{
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "document", "document1", "viewer", "user", "user1"),
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "folder", "folder1", "viewer", "user", "user1"),
+			},
+			expectedUpdates: []*v1.RelationshipUpdate{
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "document", "document1", "viewer", "user", "user1"),
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "folder", "folder1", "viewer", "user", "user1"),
+			},
+		},
+		{
+			name:              "watch with objectType filter",
+			expectedCode:      codes.OK,
+			objectTypesFilter: []string{"document"},
+			mutations: []*v1.RelationshipUpdate{
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "document", "document1", "viewer", "user", "user1"),
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "folder", "folder1", "viewer", "user", "user1"),
+			},
+			expectedUpdates: []*v1.RelationshipUpdate{
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "document", "document1", "viewer", "user", "user1"),
+			},
+		},
+		{
+			name:         "invalid zedtoken",
+			startCursor:  &v1.ZedToken{Token: "bad-token"},
+			expectedCode: codes.InvalidArgument,
+		},
+		{
+			name:         "empty zedtoken fails validation",
+			startCursor:  &v1.ZedToken{Token: ""},
+			expectedCode: codes.InvalidArgument,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			require := require.New(t)
+
+			rawDS, err := memdb.NewMemdbDatastore(0, 0, memdb.DisableGC, 0)
+			require.NoError(err)
+
+			ds, _ := testfixtures.StandardDatastoreWithSchema(rawDS, require)
+
+			revision, err := ds.SyncRevision(context.Background())
+			require.NoError(err)
+			require.True(revision.GreaterThan(decimal.Zero))
+
+			client, stop := newWatchServicer(require, ds)
+			defer stop()
+
+			cursor := zedtoken.NewFromRevision(revision)
+			if tc.startCursor != nil {
+				cursor = tc.startCursor
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			stream, err := client.Watch(ctx, &v1.WatchRequest{
+				OptionalObjectTypes: tc.objectTypesFilter,
+				OptionalStartCursor: cursor,
+			})
+			require.NoError(err)
+
+			if tc.expectedCode == codes.OK {
+
+				updatesChan := make(chan []*v1.RelationshipUpdate, len(tc.mutations))
+
+				go func() {
+					defer func() {
+						close(updatesChan)
+					}()
+
+					for {
+						select {
+						case <-ctx.Done():
+							return
+						case <-time.After(3 * time.Second):
+							panic(fmt.Errorf("timed out waiting for stream updates"))
+						default:
+							resp, err := stream.Recv()
+							if err != nil {
+								errStatus, ok := status.FromError(err)
+								if (ok && (errStatus.Code() == codes.Canceled || errStatus.Code() == codes.Unavailable)) || err == io.EOF {
+									break
+								}
+
+								panic(fmt.Errorf("received a stream read error: %v", err))
```

Comment:
`require.Fail`?

Reply:
`require.Fail` inside of a goroutine that is not the main routine that the `testing.T` object is associated with causes unclear failures such as `panic: Fail in goroutine after <process> has completed`.

Would you prefer `require.Fail` here instead?

Reply:
Ah... I guess this is okay then